src/main/java/com/ontotext/kafka/processor/SinkRecordsProcessor.java

```diff
@@ -0,0 +1,320 @@
+import org.eclipse.rdf4j.repository.http.HTTPRepository;
+import org.eclipse.rdf4j.rio.RDFFormat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.ontotext.kafka.util.ValueUtil.convertValueToString;
+
+/**
+ * A processor which batches sink records and flushes the updates to a given
+ * GraphDB {@link org.eclipse.rdf4j.repository.http.HTTPRepository}.
+ * <p>
+ * Batches that do not meet the {@link com.ontotext.kafka.GraphDBSinkConfig#BATCH_SIZE} are flushed
+ * after passing {@link com.ontotext.kafka.GraphDBSinkConfig#RECORD_POLL_TIMEOUT} threshold.
+ *
+ * @author Tomas Kovachev tomas.kovachev@ontotext.com
+ */
+public final class SinkRecordsProcessor implements Runnable {
+
+	private static final Logger LOG = LoggerFactory.getLogger(SinkRecordsProcessor.class);
+
+	private final UUID id = UUID.randomUUID();
+	private final LinkedBlockingQueue<Collection<SinkRecord>> sinkRecords;
+	private final Queue<SinkRecord> recordsBatch;
+	private final Repository repository;
+	private final RDFFormat format;
+	private final int batchSize;
+	private final long timeoutCommitMs;
+	private final LogErrorHandler errorHandler;
+	private final StringBuilder updateQueryBuilder;
+	private final String templateId;
+	private final RecordHandler recordHandler;
+	private final GraphDBSinkConfig.TransactionType transactionType;
+	private final GraphDBSinkConfig config;
+	private AtomicBoolean backOff = new AtomicBoolean(false);
+
+	public SinkRecordsProcessor(GraphDBSinkConfig config, SinkTaskContext context) {
+		this(config, new LinkedBlockingQueue<>(), initRepository(config));
+	}
+
+	public SinkRecordsProcessor(GraphDBSinkConfig config, LinkedBlockingQueue<Collection<SinkRecord>> sinkRecords, Repository repository) {
+		this.config = config;
+		this.sinkRecords = sinkRecords;
+		this.recordsBatch = new ConcurrentLinkedQueue<>();
+		this.repository = repository;
+		this.updateQueryBuilder = new StringBuilder();
+		this.templateId = config.getTemplateId();
+		this.format = config.getRdfFormat();
+		this.batchSize = config.getBatchSize();
+		this.timeoutCommitMs = config.getProcessorRecordPollTimeoutMs();
+		this.errorHandler = new LogErrorHandler(config);
+		this.transactionType = config.getTransactionType();
+		this.recordHandler = getRecordHandler(transactionType);
+		// repositoryUrl is used for logging purposes
+		String repositoryUrl = (repository instanceof HTTPRepository) ? ((HTTPRepository) repository).getRepositoryURL() : "unknown";
+		MDC.put("RepositoryURL", repositoryUrl);
+		MDC.put("Connector", config.getConnectorName());
+	}
+
+
+	private static Repository initRepository(GraphDBSinkConfig config) {
```

Comment:
What do you think about moving this in a separate class? It's logic is not directly linked to the the sink processing. This class just contains a lot of logic and might be easier to read if we move out a thing or two.